text/machine_comprehension/bert-squad/README.md

```diff
@@ -0,0 +1,84 @@
+
+## Model
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393,930 KB](model/bert.onnx)  |[MD5](BERT-md5.txt) |  [932 KB](model/saved_model.pb) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  | 
+
+## Inference
+We used onnxruntime to preform the inference. 
+
+### Input 
+Uses WordPiece tokenisation method to split the input text into list of tokens that are available in the vocabulary (30,522 words).
+Then converts these tokens into features
+<li>input_ids: list of numerical ids for the tokenised text
+<li>input_mask: will be set to 1 for real tokens and 0 for the padding tokens
+<li>segment_ids: for our case, this will be set to the list of ones
+<li>label_ids: one-hot encoded labels for the text
+
+### Preprocessing
+Get parameters and convert input examples into features
+```python
+# preprocess input 
+predict_file = 'inputs.json'
+
+# Use read_squad_examples method from run_onnx_squad to read the input file
+eval_examples = read_squad_examples(input_file=predict_file)
+
+max_seq_length = 256
+doc_stride = 128
+max_query_length = 64
+batch_size = 1
+n_best_size = 20
+max_answer_length = 30
+
+vocab_file = os.path.join('uncased_L-12_H-768_A-12', 'vocab.txt')
+tokenizer = tokenization.FullTokenizer(vocab_file=vocab_file, do_lower_case=True)
+
+# Use convert_examples_to_features method from run_onnx_squad to get parameters from the input 
+input_ids, input_mask, segment_ids, extra_data = convert_examples_to_features(eval_examples, tokenizer,
+                                                                              max_seq_length, doc_stride, max_query_length)
+```
+
+### Output
+Given a question and a context paragraph, the model predicts a start and an end token from the paragraph that most likely answers the questions. Computes a start vector and a final transformer output for each token. 
+
+### Postprocessing
+Write the predictions (answers to the questions) in a file. 
+```python
+# postprocess results
+output_dir = 'predictions'
+os.makedirs(output_dir, exist_ok=True)
+output_prediction_file = os.path.join(output_dir, "predictions.json")
+output_nbest_file = os.path.join(output_dir, "nbest_predictions.json")
+write_predictions(eval_examples, extra_data, all_results,
+                  n_best_size, max_answer_length,
+                  True, output_prediction_file, output_nbest_file)
+```
+
+## Dataset (Train and Validation)
+The model is trained with [SQuAD v1.1](https://rajpurkar.github.io/SQuAD-explorer/explore/1.1/dev/) dataset that contains 100,000+ question-answer pairs on 500+ articles. 
+
+## Validation accuracy
+Metric is Exact Matching (EM) of 84.3, computed over SQuAD v1.1 dev data.
```

Comment:
Where did this number come from?

Reply:
It is from the google-research github for the BERT uncased model because that is the model I trained with squad-1.1 dataset.

Reply:
is it possible to verify this against the ONNX model?

Reply:
I verified it against the ONNX model and updated the metric